### PR TITLE
Add quotas to Permits

### PR DESCRIPTION
**Pull request by ryanhall07**
Allows a Permit impl to decide how much quota a caller is allowed to use
for each Permit acquired. Having a quota > 1 is a nice performance
optimization to avoid having to acquire a permit everytime a resource is
consumed.

Additionally the new Permit structure allows tracking if a Permit is
acquired/released more than once by a caller using a refCount. Allows
catching bugs in tests by panicing.

**What this PR does / why we need it**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/dbnode/network/server/tchannelthrift/node/service.go [hunk 1]
```diff
@@ -939,17 +937,11 @@ type fetchTaggedResultsIterOpts struct {
 }
 
 func newFetchTaggedResultsIter(opts fetchTaggedResultsIterOpts) FetchTaggedResultsIter { //nolint: gocritic
-	_, limitBased := opts.blockPermits.(*permits.LookbackLimitPermit)
```
Comment (nbroyles) [inline src/dbnode/network/server/tchannelthrift/node/service.go:942]: So nice to be able to get rid of this

**Review by nbroyles — APPROVED** [review]

File: src/dbnode/network/server/tchannelthrift/node/service.go [hunk 2]
```diff
@@ -1038,37 +1030,45 @@ func (i *fetchTaggedResultsIter) Next(ctx context.Context) bool {
...
+	var curPermit permits.Permit
+	if len(i.permits) > 0 {
+		curPermit = i.permits[len(i.permits)-1]
+	}
+	if curPermit == nil || curPermit.QuotaRemaining() <= 0 {
 		if i.idx == idx {
 			// block acquiring if we need the block readers to fulfill the current fetch.
-			if err := i.blockPermits.Acquire(ctx); err != nil {
+			permit, err := i.blockPermits.Acquire(ctx)
+			if err != nil {
```
Comment (wesleyk) [inline src/dbnode/network/server/tchannelthrift/node/service.go:1041]: interesting; so if we do run into an error, we don't technically release all held permits? Or is that handled elsewhere? (separate comment from your change I think)

File: src/dbnode/storage/limits/permits/permit.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+// permit tracks the quota used by a caller and provides basic sanity checks that a caller
+// correctly uses the permit.
+type permit struct {
+	// immutable state
+	quota int64
+	iOpts instrument.Options
+
+	// mutable state
+	quotaUsed int64
+	refCount  atomic.Int32
```
Comment (wesleyk) [inline src/dbnode/storage/limits/permits/permit.go:39]: can we call out that this should be 0 or 1?

File: src/dbnode/storage/limits/permits/types.go [hunk 4]
```diff
@@ -43,15 +45,35 @@ type Manager interface {
...
-	// Release gives back one acquired permit from the specific permits instance. The user can pass an optional quota
-	// indicating how much of quota was used while holding the permit.
+	// Release gives back one acquired permit from the specific permits instance.
 	// Cannot release more permits than have been acquired.
-	Release(quota int64)
+	Release(permit Permit)
+}
+
+// Permit is granted to a caller which is allowed to consume some amount of quota.
+type Permit interface {
```
Comment (wesleyk) [inline src/dbnode/storage/limits/permits/types.go:62]: hmm, callers should never invoke release or acquire, fair to say? I guess the panics help guard that?

**Review by wesleyk — APPROVED** [review]
LGTM

File: src/dbnode/storage/limits/permits/types.go (hunk 4 above)
Comment (ryanhall07) [reply src/dbnode/storage/limits/permits/types.go:62]: correct. I couldn't think of a better name. maybe `PrepareForRelease` and `PrepareForAcquire`?

File: src/dbnode/network/server/tchannelthrift/node/service.go (hunk 2 above)
Comment (ryanhall07) [reply src/dbnode/network/server/tchannelthrift/node/service.go:1041]: when an error happens, the fetched tagged iterator is closed, which releases all held permits.

File: src/dbnode/storage/limits/permits/types.go (hunk 4 above)
Comment (ryanhall07) [reply src/dbnode/storage/limits/permits/types.go:62]: or maybe `PostRelease` and `PreAcquire`, kind of like hooks for the managers to call

File: src/dbnode/storage/index.go [hunk 5]
```diff
@@ -1669,23 +1668,24 @@ func (i *nsIndex) queryWithSpan(
...
 				}
 				first = false
 				startProcessing := time.Now()
-				execBlockFn(ctx, blockIter.block, blockIter.iter, opts, state, results, logFields)
+				execBlockFn(ctx, blockIter.block, permit, blockIter.iter, opts, state, results, logFields)
 				processingTime := time.Since(startProcessing)
 				queryMetrics.blockProcessingTime.RecordDuration(processingTime)
 				blockIter.processingTime += processingTime
-				permits.Release(int64(processingTime))
+				permit.Use(int64(processingTime))
```
Comment (asafm) [inline src/dbnode/storage/index.go:1683]: @ryanhall07 What's the motivation for calling .Use on the permit, if a line later it is immediately released, hence you're not using that information anywhere?

Comment (ryanhall07) [reply src/dbnode/storage/index.go:1683]: so the `Permit` can track how much of the allowed quota was actually used. impls of the `Permits` may care to know about this information.

**Pull request merged**